files/default/setup-ephemeral-drives.sh

```diff
@@ -13,124 +11,169 @@ set -x
+      DEVICES+=("/dev/${MAPPING}")
+      NUM_DEVICES=$((NUM_DEVICES + 1))
     fi
   done
-  if [ $NUM_DEVS -gt 0 ]; then
-    for d in $DEVS; do
-      d=/dev/${d}
-      dd if=/dev/zero of=${d} bs=32k count=1 || RC=1
-      exec_command "parted -s ${d} mklabel gpt"
-      exec_command "parted -s ${d}"
-      exec_command "parted -s -a optimal ${d} mkpart primary 1MB 100%"
-      partprobe ${d}
-      exec_command "parted -s ${d} set 1 lvm on"
-      if [ $IS_NVME -eq 1 ]; then
-        PARTITIONS="${d}p1 $PARTITIONS"
-      else
-        PARTITIONS="${d}1 $PARTITIONS"
-      fi
-    done
-    if [ $RC -ne 0 ]; then
-      error_exit "Failed to detect and/or partition ephemeral devices."
-    fi
 
-    # sleep 10 seconds to let partitions settle (bug?)
-    sleep 10
-
-    # Setup LVM
-    RC=0
-    pvcreate -y $PARTITIONS || RC=1
-    vgcreate vg.01 $PARTITIONS || RC=1
-    lvcreate -i $NUM_DEVS -I 64 -l 100%FREE -n lv_ephemeral vg.01 || RC=1
-    if [ "${encrypted_ephemeral}" == "true" ]; then
-      modprobe brd || RC=1
-      mkfs -q /dev/ram1 1024 || RC=1
-      mkdir -p /root/keystore || RC=1
-      mount /dev/ram1 /root/keystore || RC=1
-      dd if=/dev/urandom of=/root/keystore/keyfile bs=1024 count=4 || RC=1
-      chmod 0400 /root/keystore/keyfile || RC=1
-      cryptsetup -q luksFormat /dev/vg.01/lv_ephemeral /root/keystore/keyfile || RC=1
-      cryptsetup -d /root/keystore/keyfile luksOpen /dev/vg.01/lv_ephemeral ephemeral_luks || RC=1
-      mkfs.ext4 /dev/mapper/ephemeral_luks || RC=1
-      mount -v -t ext4 -o noatime,nodiratime /dev/mapper/ephemeral_luks ${ephemeral_dir} || RC=1
-    else
-      mkfs.ext4 /dev/vg.01/lv_ephemeral || RC=1
-      echo "/dev/vg.01/lv_ephemeral ${ephemeral_dir} ext4 noatime,nodiratime 0 0" >> /etc/fstab || RC=1
-      mount -v ${ephemeral_dir} || RC=1
-    fi
+  if [[ "${NUM_DEVICES}" -gt 0 ]]; then
+    log "This instance type has (${NUM_DEVICES}) device(s) for instance store: (${DEVICES[*]})"
+  else
+    exit_noop "This instance type doesn't have instance store"
+  fi
+
+  if [[ "${IS_NVME}" -eq 0 ]]; then
+    log "This instance store may suffer first-write penalty unless initialized: please have a look at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/disk-performance.html"
+    # Initialization can take long time, even hours
+    # for DEVICE in "${DEVICES[@]}"; do
+    #  dd if=/dev/zero of="${DEVICE}" bs=1M
+    # done
   fi
-  chmod 1777 ${ephemeral_dir} || RC=1
-  if [ $RC -ne 0 ]; then
-    error_exit "Failed to create LVM stripe and/or format ephemeral volume."
+}
+
+function create_lvm {
+  log "Creating LVM (${LVM_PATH})"
+  pvcreate -y "${DEVICES[@]}"
+  vgcreate -y vg.01 "${DEVICES[@]}"
+  lvcreate -y -i ${NUM_DEVICES} -I 64 -l 100%FREE -n "${LVM_NAME}" "${LVM_VG_NAME}"
+}
+
+function check_lvm_exist {
+  LVM_EXIST_COMMAND="lvs ${LVM_PATH} --nosuffix --noheadings -q"
+
+  if ! ${LVM_EXIST_COMMAND} &>/dev/null; then
+    log "LVM (${LVM_PATH}) does not exist"
+    create_lvm
```

Comment:
shall we check if this command exits with 0?

Reply:
done